### PR TITLE
Allow support users to remove claims

### PR DESCRIPTION
**Pull request by gms-gs**
## Context

We need to allow the support user to remove claims. Currently the only states the claim can be in are draft and submitted, at least user facing.

## Changes proposed in this pull request

Implement the discarded gem for a Claim

## Guidance to review

- Creare a draft and submitted claim
- View the claim
- Follow the deleted flow
- Claim will not appear 
- Claim will be discarded

## Link to Trello card

https://trello.com/c/JUhoXxIJ/259-allow-support-users-to-remove-claims

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated or [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/models/claim.rb [hunk 1]
```diff
@@ -30,17 +30,17 @@ class Claim < ApplicationRecord
...
   validates :status, presence: true
   validates :reference, uniqueness: { case_sensitive: false }, allow_nil: true
 
-  scope :not_internal, -> { where.not(status: :internal) }
+  scope :not_internal_or_discarded, -> { where.not(status: %i[internal discarded]) }
   scope :order_created_at_desc, -> { order(created_at: :desc) }
 
   enum :status,
-       { internal: "internal", draft: "draft", submitted: "submitted" },
+       { internal: "internal", draft: "draft", submitted: "submitted", discarded: "discarded" },
```
Comment (JamieCleare2525) [inline app/models/claim.rb:43]: Why aren't you using the `discard` gem for making a record discarded? We already have this functionality for `Users`, and should keep the discarding mechanism consistent throughout the service/application

Comment (JamieCleare2525) [reply app/models/claim.rb:43]: A key feature of using `discarded` is that, should you need to `undiscard` the record you haven't lost it's original state.

File: app/models/claim.rb [hunk 2]
```diff
@@ -30,17 +30,17 @@ class Claim < ApplicationRecord
...
 
-  has_many :mentor_trainings
+  has_many :mentor_trainings, dependent: :destroy
   has_many :mentors, through: :mentor_trainings
 
   validates :status, presence: true
   validates :reference, uniqueness: { case_sensitive: false }, allow_nil: true
 
-  scope :not_internal, -> { where.not(status: :internal) }
+  scope :not_internal_or_discarded, -> { where.not(status: %i[internal discarded]) }
```
Comment (JamieCleare2525) [inline app/models/claim.rb:39]: Might be better to have an `active` scope, rather than this very specifically named scope.
Could be beneficially to have a constant (e.g `ACTIVE_STATUSES = %i[draft submitted].freeze`), rather than doing the negative of that search. 
```suggestion
  scope :active, -> { where(status: ACTIVE_STATUSES) }
```

File: app/controllers/claims/support/schools/claims_controller.rb [hunk 3]
```diff
@@ -1,18 +1,27 @@
...
 
   def new; end
 
   def show; end
 
+  def remove; end
+
+  def destroy
+    @claim.destroy! if @claim.draft?
+    discard_submitted_claim if @claim.submitted?
```
Comment (JamieCleare2525) [inline app/controllers/claims/support/schools/claims_controller.rb:20]: Is there a case where the claim would not be either `draft` or `submitted`?
```suggestion
    @claim.submitted? ? @claim.discard : @claim.destroy!
```

File: app/policies/claim_policy.rb [hunk 4]
```diff
@@ -11,6 +11,14 @@ def submit?
     !user.support_user? && !record.submitted?
   end
 
+  def destroy?
+    user.support_user?
```
Comment (JamieCleare2525) [inline app/policies/claim_policy.rb:15]: You could additionally ensure that the claim is in one of the states which will allow you do perform this action.
```suggestion
    user.support_user? && (record.submitted? || record.draft?)
```

File: app/controllers/claims/support/schools/claims_controller.rb (hunk 3 above)
Comment (ollietreend) [reply app/controllers/claims/support/schools/claims_controller.rb:20]: I'm wondering if it'd be worth using soft-deletion everywhere for the Claim model?

It feels a bit unusual to implement soft-delete only for 'submitted' claims, and instead to permanently delete 'draft' claims. I could imagine that being confusing behaviour – for example, if somebody accidentally deletes a 'submitted' claim, we'd be able to help them retrieve it by un-discarding it. But the same wouldn't be true if the claim happened to have been in a 'draft' state, since it would have been permanently deleted.

This is more of a data modelling decision rather than an implementation detail, so I'm wondering if @Nitemaeric has an opinion? It's the sort of thing that's worth getting right upfront.

And as @JamieCleare2525 said, we implemented the [discard](https://github.com/jhawthorn/discard) gem back in PR #222, so it'd be good to follow that approach as a consistent way to soft-delete records within the app.

Comment (Nitemaeric) [reply app/controllers/claims/support/schools/claims_controller.rb]: This is a Product requirement we've received, https://trello.com/c/JUhoXxIJ/259-allow-support-users-to-remove-claims#comment-65fbedab17e1be89451976e0. We've decided that "archived" is equal to "soft-deleted" and opted towards using the `discard` gem to manage that.

The support interface will have a way to view "archived" records, so we want to make the distinction between hard-deleting drafts and soft-deleting (archiving) submitted claims.

We _may_ be incorrectly equating "archived" and "soft-deleted". This is open for debate.

File: app/models/claim.rb [hunk 5]
```diff
@@ -30,13 +30,15 @@ class Claim < ApplicationRecord
   belongs_to :provider
   belongs_to :created_by, polymorphic: true
 
-  has_many :mentor_trainings
+  has_many :mentor_trainings, dependent: :destroy
```
Comment (Nitemaeric) [inline app/models/claim.rb:33]: Let's add the `shoulda-matchers` modifier, `.dependent`, in the model specs.

File: app/models/claims/school.rb [hunk 6]
```diff
@@ -14,9 +14,9 @@
 #  gender                       :string
 #  group                        :string
 #  last_inspection_date         :date
+#  latitude                     :float
 #  local_authority_code         :string
 #  local_authority_name         :string
-#  latitude                     :float
 #  longitude                    :float
```
Comment (Nitemaeric) [inline app/models/claims/school.rb:20]: This looks unrelated to the PR scope.

File: app/models/placements/school.rb [hunk 7]
```diff
@@ -14,9 +14,9 @@
 #  gender                       :string
 #  group                        :string
 #  last_inspection_date         :date
+#  latitude                     :float
 #  local_authority_code         :string
 #  local_authority_name         :string
-#  latitude                     :float
```
Comment (Nitemaeric) [inline app/models/placements/school.rb:19]: This looks unrelated to the PR scope.

File: app/models/school.rb [hunk 8]
```diff
@@ -14,9 +14,9 @@
 #  gender                       :string
 #  group                        :string
 #  last_inspection_date         :date
+#  latitude                     :float
 #  local_authority_code         :string
 #  local_authority_name         :string
-#  latitude                     :float
```
Comment (Nitemaeric) [inline app/models/school.rb:19]: This looks unrelated to the PR scope.

File: app/policies/claim_policy.rb [hunk 9]
```diff
@@ -11,6 +11,14 @@ def submit?
...
   end
 
+  def destroy?
+    user.support_user? && record.draft?
+  end
+
+  def remove?
+    destroy?
+  end
+
```
Comment (Nitemaeric) [inline app/policies/claim_policy.rb:21]: This is already in `Claims::ApplicationPolicy`.

File: app/views/claims/support/schools/claims/show.html.erb [hunk 10]
```diff
@@ -71,4 +71,8 @@
         <% end %>
     </div>
   </div>
+
+  <% if @claim.draft? %>
```
Comment (Nitemaeric) [inline app/views/claims/support/schools/claims/show.html.erb:75]: We should rely on the policy so we don't need to repeat ourselves.

```suggestion
  <% if policy(@claim).remove? %>
```

File: spec/factories/schools.rb [hunk 11]
```diff
@@ -14,9 +14,9 @@
 #  gender                       :string
 #  group                        :string
 #  last_inspection_date         :date
+#  latitude                     :float
 #  local_authority_code         :string
 #  local_authority_name         :string
-#  latitude                     :float
```
Comment (Nitemaeric) [inline spec/factories/schools.rb:19]: This looks unrelated to the PR scope.

File: spec/factories/subjects.rb [hunk 12]
```diff
@@ -1,3 +1,14 @@
...
+# Table name: subjects
+#
+#  id           :uuid             not null, primary key
+#  code         :string
+#  name         :string           not null
+#  subject_area :enum
+#  created_at   :datetime         not null
+#  updated_at   :datetime         not null
+#
 FactoryBot.define do
```
Comment (Nitemaeric) [inline spec/factories/subjects.rb:12]: This looks unrelated to the PR scope.

File: spec/models/claims/school_spec.rb [hunk 13]
```diff
@@ -14,9 +14,9 @@
 #  gender                       :string
 #  group                        :string
 #  last_inspection_date         :date
+#  latitude                     :float
 #  local_authority_code         :string
 #  local_authority_name         :string
-#  latitude                     :float
```
Comment (Nitemaeric) [inline spec/models/claims/school_spec.rb:19]: This looks unrelated to the PR scope.

File: spec/models/placements/school_spec.rb [hunk 14]
```diff
@@ -14,9 +14,9 @@
 #  gender                       :string
 #  group                        :string
 #  last_inspection_date         :date
+#  latitude                     :float
 #  local_authority_code         :string
 #  local_authority_name         :string
-#  latitude                     :float
```
Comment (Nitemaeric) [inline spec/models/placements/school_spec.rb:19]: This looks unrelated to the PR scope.

File: spec/models/school_spec.rb [hunk 15]
```diff
@@ -14,9 +14,9 @@
 #  gender                       :string
 #  group                        :string
 #  last_inspection_date         :date
+#  latitude                     :float
 #  local_authority_code         :string
 #  local_authority_name         :string
-#  latitude                     :float
```
Comment (Nitemaeric) [inline spec/models/school_spec.rb:19]: This looks unrelated to the PR scope.

**Review by Nitemaeric — COMMENTED** [review]
There is quite a bit of unrelated changes. Let's remove them from the PR - or - put them in their own commit.

**Review by Nitemaeric — APPROVED** [review]

File: spec/models/claim_spec.rb [hunk 16]
```diff
@@ -59,13 +59,13 @@
   end
 
   describe "scopes" do
-    describe "#not_internal" do
+    describe "#visible" do
       it "returns the claims that dont have status internal" do
         create(:claim)
         claim1 = create(:claim, :draft)
         claim2 = create(:claim, :submitted)
```
Comment (JamieCleare2525) [inline spec/models/claim_spec.rb:66]: Can you add claims to this test which aren't `:draft` or `:submitted`. As currently it is returning `Claim.all`

File: spec/system/claims/support/schools/claims/removing_a_claim_spec.rb [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+  def when_i_click_remove_claim
+    click_on "Remove claim"
+  end
+
+  def when_i_confirm_removal
+    expect(page).to have_content("Are you sure you want to remove this claim?")
+    click_on "Remove claim"
+  end
+
+  def then_the_claim_has_been_deleted
```
Comment (JamieCleare2525) [inline spec/system/claims/support/schools/claims/removing_a_claim_spec.rb:91]: Could you also check for the flash message, to:
1. Show that the correct controller action has occurred.
2. Have confirmation that the deletion has happened that that we aren't just looking at a page without that content.

**Review by JamieCleare2525 — APPROVED** [review]
